scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/StepSimulation.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.commons.lang3.tuple.Triple;
+
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+public class StepSimulation {
+
+  private Duration curTime;
+  private SimulationEngine engine;
+  private LiveCells cells;
+  private TemporalEventSource timeline;
+  private MissionModel<?> missionModel;
+
+  //used for looking up durations
+  private Map<SerializedActivity, TaskId> actToId;
+
+  //boolean stating whether the simulation results should be regenerated
+  private boolean areLastSimResultsDirty;
+
+  //simulation results so far
+  private SimulationResults lastSimResults;
+
+  //should the simulation be resetted everytime an activity is simulated
+  private boolean resetEveryTime;
+
+  //map from activity start time to serialized activity and its name
+  private LinkedHashMap<Duration, Pair<SerializedActivity, String>> activitiesInserted;
+
+  /*when we really want to reset the whole simulation*/
+  public void clearActInserted(){
+    activitiesInserted.clear();
+  }
+
+  public StepSimulation(MissionModel<?> missionModel){
+    this.missionModel = missionModel;
+    resetEveryTime = false;
+    activitiesInserted = new LinkedHashMap<>();
+  }
+
+  public void resetEveryTime(){
+    resetEveryTime= true;
+  }
+
+  public void initSimulation(){
+    actToId = new HashMap<>();
+    lastSimResults = null;
+    this.engine = new SimulationEngine();
+    activitiesInserted.clear();
+
+      /* The top-level simulation timeline. */
+      this.timeline = new TemporalEventSource();
+      areLastSimResultsDirty = false;
+      this.cells = new LiveCells(timeline, missionModel.getInitialCells());
+      /* The current real time. */
+      curTime = Duration.ZERO;
+
+      // Begin tracking all resources.
+      for (final var entry : missionModel.getResources().entrySet()) {
+        final var name = entry.getKey();
+        final var resource = entry.getValue();
+        engine.trackResource(name, resource, curTime);
+      }
+
+      // Start daemon task(s) immediately, before anything else happens.
+      {
+        final var daemon = engine.initiateTaskFromSource(missionModel::getDaemon);
+        final var commit = engine.performJobs(Set.of(SimulationEngine.JobId.forTask(daemon)),
+                                              cells, curTime, Duration.MAX_VALUE, missionModel);
+        timeline.add(commit);
+      }
+  }
+
+  public void simulateActivity(SerializedActivity activity, Duration startTime, String name){
+    if(startTime.shorterThan(curTime) || resetEveryTime){
+      var toBeReinserted = new LinkedHashMap<Duration, Pair<SerializedActivity, String>>();
+      toBeReinserted.putAll(activitiesInserted);
+      initSimulation();
+      insertAllActsBefore(toBeReinserted, startTime);
+      simulateAct(activity, startTime,name);
```

Comment:
Here and a couple lines below, missing whitespace for consistency:
```suggestion
      simulateAct(activity, startTime, name);
```